make telegram handler persist chats/groups across sessions, fix json serialization bug, fix key error

fried-sausage opened the pull request:
Implements #130 
Fixes #127 
Fixes #129 
Some methods return bytes on python3 which was reason of TypeError during json serialization.
Simply using bytes.decode() was enough.
Also removed str.encode() method, which returns bytes on python3 and fixed key error (sometimes handler receives update without 'text' field).

garrettheel commented:
Would you mind rebasing this off master?

fried-sausage commented:
@garrettheel ok, will do shortly

fried-sausage commented:
@garrettheel i believe it`s done, is everything ok?

garrettheel commented:
@fried-sausage there are currently conflicts and a bunch of commits contained on this branch that are not relevant (it makes it very difficult to read).

Can you rebase off develop and ensure that only your commits show up on the "Commits" tab of this PR?

fried-sausage commented:
@garrettheel ok, only my commits now. Sorry for sloppiness

garrettheel commented on README.md:
```diff
@@ -404,7 +404,8 @@ Sends a Telegram message.
         "token": "telegram bot token",
-        "bot_ident": "token used to activate bot in a group"
+        "bot_ident": "token you choose to activate bot in a group"
+        "chatfile": "path to file where chat id`s are saved"
```
Can you make it clear that this is optional and change `id<backtick>s` => `ids`?

garrettheel commented on graphite_beacon/handlers/telegram.py:
```diff
@@ -23,89 +29,149 @@ def init_handler(self):
         assert self.bot_ident, 'Telegram bot ident token is not defined.'
 
+        self.chatfile = self.options.get('chatfile')
+        if not self.chatfile: LOGGER.warning(NO_CHATFILE)
```
Please use a new indented line for these blocks

garrettheel commented on graphite_beacon/handlers/telegram.py:
```diff
@@ -23,89 +29,149 @@ def init_handler(self):
+            return []
+
+    def _prepare_request(self, target, body, method='POST',
+                            headers={"Content-Type": "application/json"}):
```
indentation is wrong here

garrettheel commented on graphite_beacon/handlers/telegram.py:
```diff
@@ -23,89 +29,149 @@ def init_handler(self):
+            update_id = update.get('update_id')
+            if not update_id: continue
+
+            message = update.get('message')  # fix key error here
```
Don't need this comment anymore :)

garrettheel commented on graphite_beacon/handlers/telegram.py:
```diff
@@ -23,89 +29,149 @@ def init_handler(self):
+            chat_id = message.get('chat', {}).get('id')
+            if not chat_id: continue
+
+            if not self._correct_activation(text, chat_id):
```
Can you fetch everything that is required and do the `continue` with one `if` statement?
E.g
```python
if not all(update_id, message, text, message_id, ...):
  continue
```

Also is this an exceptional circumstance worth logging?

fried-sausage replied:
> Also is this an exceptional circumstance worth logging?

Well, the absence of `message` field in `update` is normal (and not worth logging IMO) -- it happens when someone edits message in chat for example.

`_correct_activation` method is a whole different story, generally also not worth logging. Imagine chat conference (aka telegram group) with several participants including our bot, we would be logging every time someone posts. But we might consider logging when user tries to activate bot with wrong bot_ident. If someone thinks this case should be logged ill do it (should it be Logger.debug/Logger.warning/Logger.error?).

The rest of cases are not so clear to me, not sure whether all those situations are even possible. With exception of `if not update_id: continue` all those continue guards already were in code, i just put them all in one place one after another.
Will look closer into Telegram API, but im quite sure every update has update_id, so at least this `continue` is not needed.

fried-sausage replied:
> Can you fetch everything that is required and do the continue with one if statement?

Yes, thanks for remark, i will put it all in one function (although `all(iterable)` is not fit because we need to implement logs inside and we are forced to do some additional steps with it).

garrettheel replied:
Sounds good in regards to error handling, how about `warning` on activation with the wrong bot_ident?

garrettheel commented on graphite_beacon/handlers/telegram.py:
```diff
@@ -23,89 +29,149 @@ def init_handler(self):
+            with open(self.chatfile) as fh:
+                return [int(chat) for chat in fh]
+        except Exception as e:
+            LOGGER.debug('could not load saved chats:\n%s' % (e))
```
This should be logged as an error

garrettheel commented:
@fried-sausage I left some comments, but generally looks good. I'm not too familiar with telegram so some more comments around how things work would be greatly appreciated.

Do chat IDs written to the file ever expire and need to be removed?

fried-sausage commented:
@garrettheel your remark about chat ID's is very timely actually.
ID's do not expire, but we need a way to deactivate bot in chat,  lets say `/deactivate` command, and when we do so, we better remove corresponding chat ID from our file.
I was going to implement this, will do soon.
I will try to add comments and docstrings, although my understanding of telegram API might be superficial.

fried-sausage commented:
@garrettheel please see, let me know if you think something should be done differently.
Added /help and /deactivate commands.
Added getUpdates and sendMessage methods named after corresponding Telegram api methods. We can now add other methods, for example sendDocument, by giving method name string to _fetchmaker.
Chats are now stored in set instead of list (we frequently check membership but dont do indexing at all).

garrettheel commented on graphite_beacon/handlers/telegram.py:
```diff
@@ -23,89 +44,173 @@ def init_handler(self):
-        self.url = "https://api.telegram.org/bot%s/" % (self.token)
-
-        self._chats = []
+        file = self.options.get('chatfile')
```
`file` is a builtin, please use a different variable name

fried-sausage replied:
check

garrettheel commented on graphite_beacon/handlers/telegram.py:
```diff
@@ -23,89 +44,173 @@ def init_handler(self):
+        data = get_data(upd, self.bot_ident)
+        for (update_id, chat_id, message_id, command) in data:
+            self._last_update = update_id
+            chat_is_known = (chat_id in chats)
```
Parens aren't needed here

garrettheel commented on graphite_beacon/handlers/telegram.py:
```diff
@@ -23,89 +44,173 @@ def init_handler(self):
+
+def write_to_file(chats, chatfile):
+    with open(chatfile, 'w') as fh:
+        chats = map(str, chats)
```
What's the utility of this? Looks like it just turns a set into a list, but sets are iterable so this shouldn't be necessary

garrettheel replied:
Ah nevermind, I see that they're ints. Maybe `fh.write('\n'.join([str(id) for id in chats]))` is cleaner?

fried-sausage replied:
check

garrettheel commented on graphite_beacon/handlers/telegram.py:
```diff
@@ -23,89 +44,173 @@ def init_handler(self):
+    given to our bot.
+    """
+    is_group = (chat_id < 0) # always negative for groups
+    splitted = text.split()[:2]
```
grammar: can you call this `split_cmd` or similar?

garrettheel commented on graphite_beacon/handlers/telegram.py:
```diff
@@ -23,89 +44,173 @@ def init_handler(self):
+    if not command:
+        return
+    return (upd['update_id'], chat_id, msg['message_id'], command)
+
```
Do you mind running pylint on this file? There are some small things like 2 newlines between functions that would be nice to see :)

garrettheel commented:
@fried-sausage thanks again for all of the fixes! One last round of comments with some small nits, but happy to merge after that. I'll make a new release soon after too

fried-sausage commented:
@garrettheel travis tests fail at pep8/pylint.
for example
ERROR: InvocationError: '/home/travis/build/klen/graphite-beacon/.tox/pylint/bin/pylint --reports no graphite_beacon'
i have both installed locally and written in travis.yml file
Is this some error on my side?

garrettheel commented:
@fried-sausage tests are failing legitimately :) Can you fix the `telegram.py` ones?

You can see the reasons if you look at the travis console logs:

```
--pylint--
************* Module graphite_beacon.handlers.telegram
C: 60, 8: Invalid attribute name "getUpdates" (invalid-name)
C: 61, 8: Invalid attribute name "sendMessage" (invalid-name)
R: 91, 4: Too many branches (13/12) (too-many-branches)
W:151,45: Unused argument 'rule' (unused-argument)
W: 79, 8: Attribute '_last_update' defined outside __init__ (attribute-defined-outside-init)
W:103,12: Attribute '_last_update' defined outside __init__ (attribute-defined-outside-init)

--pep8--
graphite_beacon/handlers/telegram.py:139:21: E123 closing bracket does not match indentation of opening bracket's line
```

fried-sausage commented:
@garrettheel 
> W: 74, 8: Attribute '_last_update' defined outside __init__ (attribute-defined-outside-init)

Could you look a bit closer to this?
As i get i get it, those attibutes actually **need** to be defined outside of `__init__` with our current architecture.
They all must be defined in `init_handler` method, because init_handler is [called](https://github.com/klen/graphite-beacon/blob/develop/graphite_beacon/handlers/__init__.py#L42) from AbstractHandler's `__init__`.
Other than that, pep8 are now ok with the code.

garrettheel commented on graphite_beacon/handlers/untitled.py:
```diff
@@ -0,0 +1 @@
+make file saving static method
```
You probably want to delete this :)

garrettheel commented:
No worries, I'll turn that specific pylint check off. Thanks for fixing everything, I'll merge once that extra file is deleted

fried-sausage commented:
@garrettheel done
